src/main/java/org/wildfly/security/password/spec/DefaultPasswordSpecEncoding.java

```diff
@@ -0,0 +1,233 @@
+     * @return a byte array representing the encoded password or null if no encoder was capable to encode the given password
+     */
+    public static byte[] encode(Password password) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        PasswordFactory passwordFactory = PasswordFactory.getInstance(password.getAlgorithm());
+
+        if (ClearPassword.class.isInstance(password)) {
+            return encodeClearPasswordSpec(password, passwordFactory);
+        } else if (DigestPassword.class.isInstance(password)) {
+            return encodeDigestPasswordSpec(password, passwordFactory);
+        } else {
+            try {
+                return encodeSaltedHashPasswordSpec(password, passwordFactory);
+            } catch (InvalidKeySpecException ignore) {}
+            try {
+                return encodeIteratedSaltedHashSpec(password, passwordFactory);
+            } catch (InvalidKeySpecException ignore) {}
+            try {
+                return encodeHashPasswordSpec(password, passwordFactory);
+            } catch (InvalidKeySpecException ignore) {}
+        }
+
+        return null;
+    }
+
+    /**
+     * Decode the given byte array and create a {@link PasswordSpec} from it.
+     *
+     * @param encoded the byte array representing the encoded password
+     * @return a {@link PasswordSpec} instance created from the encoded password or null if no decoder was capable to decode the given format.
+     */
+    public static PasswordSpec decode(byte[] encoded) {
+        ByteBuffer buffer = ByteBuffer.wrap(encoded);
+        byte identifier;
+
+        try {
+            identifier = buffer.get();
+        } catch (Exception e) {
+            throw ElytronMessages.log.couldNotObtainKeySpecEncodingIdentifier();
+        }
+
+        switch (identifier) {
+            case CLEAR_PASSWORD_SPEC_ID:
+                return decodeClearPasswordSpec(buffer);
+            case DIGEST_PASSWORD_SPEC_ID:
+                return decodeDigestPasswordSpec(buffer);
+            case HASH_PASSWORD_SPEC_ID:
+                return decodeHashPasswordSpec(buffer);
+            case SALTED_HASH_PASSWORD_SPEC_ID:
+                return decodeSaltedHashPasswordSpec(buffer);
+            case ITERATED_SALTED_HASH_SPEC_ID:
+                return decodeIteratedSaltedHashPasswordSpec(buffer);
+            default:
+                return null;
+        }
+    }
+
+    private static byte[] encodeIteratedSaltedHashSpec(Password password, PasswordFactory passwordFactory) throws InvalidKeySpecException {
+        IteratedSaltedHashPasswordSpec keySpec = passwordFactory.getKeySpec(password, IteratedSaltedHashPasswordSpec.class);
+        byte[] hash1 = keySpec.getHash();
+        ByteBuffer specBytes = ByteBuffer.allocate(1 + 4 + SALT_SIZE + hash1.length)
+                .put(ITERATED_SALTED_HASH_SPEC_ID) // 1 byte identifying the password spec
+                .putInt(keySpec.getIterationCount()); // 4 bytes representing the iteration count
```

Comment:
Hummm ... I can try it out. Didn't notice it.

Regarding packed integers, should we go to this path then ? And pack salt and iteration count in a single integer ?